OpenRCT2 won't build with GCC 14

### Operating System

Arch Linux

### OpenRCT2 build

0.4.11.r77.g6df16c7946-1

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

This is a development issue

### Describe the issue

When GCC14 is installed and used to compile OpenRCT2, the code will not build.

It will throw errors like:

<pre>/usr/bin/c++ -DDOCDIR=\"/usr/share/doc/openrct2\" -DOPENGL_NO_LINK -D_FILE_OFFSET_BITS=64 -D__ENABLE_DISCORD__ -I/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/DISCORDRPC_PROCESS_INCLUDES -I/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/../thirdparty/duktape -I/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/libopenrct2 -isystem /usr/include/freetype2 -isystem /usr/include/libpng16 -isystem /usr/include/harfbuzz -isystem /usr/include/glib-2.0 -isystem /usr/lib/glib-2.0/include -isystem /usr/include/sysprof-6 -isystem /home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/../thirdparty -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -Wno-error=maybe-uninitialized  -Wno-error=restrict  -Wno-error=null-dereference  -Wno-error=stringop-overflow -fstrict-overflow -fstrict-aliasing -Werror -Wundef -Wmissing-declarations -Winit-self -Wall -Wextra -Wshadow -Wno-unknown-pragmas -Wno-missing-braces -Wno-comment -Wnonnull -Wno-unused-parameter -Wno-attributes -DDEBUG=0 -fno-char8_t -Wno-deprecated-declarations -O2 -g -DNDEBUG -std=gnu++20 -fPIC -Wsuggest-override -Wduplicated-cond -Wnon-virtual-dtor -Wduplicated-branches -Wrestrict -Wmissing-field-initializers -Wlogical-op -Wold-style-cast -Wunused-const-variable=1 -Wno-clobbered -Wredundant-decls -Wnull-dereference -Wignored-qualifiers -Wstrict-overflow=1 -DENABLE_SCRIPTING -pthread -MD -MT CMakeFiles/libopenrct2.dir/src/openrct2/actions/GameActionRegistry.cpp.o -MF CMakeFiles/libopenrct2.dir/src/openrct2/actions/GameActionRegistry.cpp.o.d -o CMakeFiles/libopenrct2.dir/src/openrct2/actions/GameActionRegistry.cpp.o -c /home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/GameActionRegistry.cpp
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:76:33: error: no matching function for call to ‘max(<brace-enclosed initializer list>)’
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:76:114: error: template argument 2 is invalid
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:76:10: error: ‘<expression error>’ in namespace ‘std’ does not name a type
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:81:52: error: ‘data’ was not declared in this scope
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:87:60: error: ‘data’ was not declared in this scope
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:93:58: error: ‘data’ was not declared in this scope
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:98:34: error: static assertion failed
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:99:34: error: static assertion failed
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:100:34: error: static assertion failed
FAILED: CMakeFiles/libopenrct2.dir/src/openrct2/actions/LandSmoothAction.cpp.o 
/usr/bin/c++ -DDOCDIR=\"/usr/share/doc/openrct2\" -DOPENGL_NO_LINK -D_FILE_OFFSET_BITS=64 -D__ENABLE_DISCORD__ -I/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/DISCORDRPC_PROCESS_INCLUDES -I/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/../thirdparty/duktape -I/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/libopenrct2 -isystem /usr/include/freetype2 -isystem /usr/include/libpng16 -isystem /usr/include/harfbuzz -isystem /usr/include/glib-2.0 -isystem /usr/lib/glib-2.0/include -isystem /usr/include/sysprof-6 -isystem /home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/../thirdparty -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -Wno-error=maybe-uninitialized  -Wno-error=restrict  -Wno-error=null-dereference  -Wno-error=stringop-overflow -fstrict-overflow -fstrict-aliasing -Werror -Wundef -Wmissing-declarations -Winit-self -Wall -Wextra -Wshadow -Wno-unknown-pragmas -Wno-missing-braces -Wno-comment -Wnonnull -Wno-unused-parameter -Wno-attributes -DDEBUG=0 -fno-char8_t -Wno-deprecated-declarations -O2 -g -DNDEBUG -std=gnu++20 -fPIC -Wsuggest-override -Wduplicated-cond -Wnon-virtual-dtor -Wduplicated-branches -Wrestrict -Wmissing-field-initializers -Wlogical-op -Wold-style-cast -Wunused-const-variable=1 -Wno-clobbered -Wredundant-decls -Wnull-dereference -Wignored-qualifiers -Wstrict-overflow=1 -DENABLE_SCRIPTING -pthread -MD -MT CMakeFiles/libopenrct2.dir/src/openrct2/actions/LandSmoothAction.cpp.o -MF CMakeFiles/libopenrct2.dir/src/openrct2/actions/LandSmoothAction.cpp.o.d -o CMakeFiles/libopenrct2.dir/src/openrct2/actions/LandSmoothAction.cpp.o -c /home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/LandSmoothAction.cpp
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:76:33: error: no matching function for call to ‘max(<brace-enclosed initializer list>)’
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:76:114: error: template argument 2 is invalid
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:76:10: error: ‘<expression error>’ in namespace ‘std’ does not name a type
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:81:52: error: ‘data’ was not declared in this scope
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:87:60: error: ‘data’ was not declared in this scope
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:93:58: error: ‘data’ was not declared in this scope
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:98:34: error: static assertion failed
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:99:34: error: static assertion failed
/home/solarisfire/.cache/yay/openrct2-git/src/openrct2-git/src/openrct2/actions/../ride/../paint/Paint.h:100:34: error: static assertion failed
ninja: build stopped: subcommand failed.
==> ERROR: A failure occurred in build().</pre> 

### Steps to reproduce

Install GCC14.
Try and build OpenRCT2.

### Attachments

_No response_

--- a/src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp
+++ b/src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp
@@ -22,6 +22,7 @@
 #    include "TransparencyDepth.h"
 
 #    include <SDL.h>
+#    include <algorithm>
 #    include <cmath>
 #    include <openrct2-ui/interface/Window.h>
 #    include <openrct2/config/Config.h>

--- a/src/openrct2-ui/windows/ShortcutKeys.cpp
+++ b/src/openrct2-ui/windows/ShortcutKeys.cpp
@@ -10,6 +10,7 @@
 #include "../input/ShortcutManager.h"
 #include "Window.h"
 
+#include <algorithm>
 #include <openrct2-ui/interface/Widget.h>
 #include <openrct2/drawing/Drawing.h>
 #include <openrct2/localisation/Formatter.h>

--- a/src/openrct2/core/Algorithm.hpp
+++ b/src/openrct2/core/Algorithm.hpp
@@ -9,6 +9,7 @@
 
 #pragma once
 
+#include <algorithm>
 #include <functional>
 
 template<class ForwardIt, class T, class Compare = std::less<>>

--- a/src/openrct2/core/Imaging.cpp
+++ b/src/openrct2/core/Imaging.cpp
@@ -19,6 +19,7 @@
 #include "Memory.hpp"
 #include "String.hpp"
 
+#include <algorithm>
 #include <fstream>
 #include <png.h>
 #include <stdexcept>

--- a/src/openrct2/drawing/X8DrawingEngine.cpp
+++ b/src/openrct2/drawing/X8DrawingEngine.cpp
@@ -24,6 +24,7 @@
 #include "LightFX.h"
 #include "Weather.h"
 
+#include <algorithm>
 #include <cstring>
 
 using namespace OpenRCT2;

--- a/src/openrct2/entity/EntityTweener.cpp
+++ b/src/openrct2/entity/EntityTweener.cpp
@@ -14,6 +14,7 @@
 #include "EntityList.h"
 #include "EntityRegistry.h"
 
+#include <algorithm>
 #include <cmath>
 
 void EntityTweener::AddEntity(EntityBase* entity)

--- a/src/openrct2/object/Object.cpp
+++ b/src/openrct2/object/Object.cpp
@@ -23,6 +23,7 @@
 #include "ObjectLimits.h"
 #include "ObjectRepository.h"
 
+#include <algorithm>
 #include <cstring>
 #include <stdexcept>
 

--- a/src/openrct2/object/ObjectTypes.cpp
+++ b/src/openrct2/object/ObjectTypes.cpp
@@ -12,6 +12,8 @@
 #include "../util/Util.h"
 #include "Object.h"
 
+#include <algorithm>
+
 constexpr std::array kAllObjectTypes = {
     ObjectType::Ride,
     ObjectType::SmallScenery,

--- a/src/openrct2/paint/Paint.h
+++ b/src/openrct2/paint/Paint.h
@@ -73,7 +73,10 @@ struct PaintStringStruct
 struct PaintEntry
 {
 private:
-    std::array<uint8_t, std::max({ sizeof(PaintStruct), sizeof(AttachedPaintStruct), sizeof(PaintStringStruct) })> data;
+    // Avoid including expensive <algorithm> for std::max. Manually ensure we use the largest type.
+    static_assert(sizeof(PaintStruct) >= sizeof(AttachedPaintStruct));
+    static_assert(sizeof(PaintStruct) >= sizeof(PaintStringStruct));
+    std::array<uint8_t, sizeof(PaintStruct)> data;
 
 public:
     PaintStruct* AsBasic()

--- a/src/openrct2/util/Util.cpp
+++ b/src/openrct2/util/Util.cpp
@@ -19,6 +19,7 @@
 #include "../scenes/title/TitleScene.h"
 #include "zlib.h"
 
+#include <algorithm>
 #include <cctype>
 #include <cmath>
 #include <ctime>

